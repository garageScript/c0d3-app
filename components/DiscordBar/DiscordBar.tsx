import React, { useEffect, useState } from 'react'
import styles from './discordBar.module.scss'
type Member = {
  id: string
  username: string
  status: string
  avatar_url: string
}
interface discordJSON {
  members: Member[]
}
interface State extends discordJSON {
  error?: string
}
const TWO_MINUTES = 2 * 60 * 1000
const Member: React.FC<{ avatar: string }> = ({ avatar }) => {
  return (
    <div className="d-inline-block">
      <img src={avatar} className={`${styles['member-avatar']} mx-1`} />
    </div>
  )
}
const DiscordBar: React.FC = () => {
  const [data, setData] = useState<State>({
    members: []
  })
  const getData = async () => {
    try {
      const data = await fetch(
        'https://discord.com/api/guilds/828783458469675019/widget.json'
      )
      const json = await data.json()
      if ('members' in json) setData(json as discordJSON)
      else setData({ members: [], error: 'Something went wrong ;(' })
    } catch (e) {
      setData({
        members: [],
        error: `Error fetching data...`
      })
    } finally {
      setTimeout(getData, TWO_MINUTES)
    }
  }
  useEffect(() => {
    getData()
  }, [])
  const reshuffle = (data: discordJSON) => {
    //randomize user order and select first five avatars to render
    const shuffled = data.members.sort(() => 0.5 - Math.random())
    const remaining = data.members.length - 5
    const profiles = shuffled.slice(0, 5).map(member => {
      return <Member avatar={member.avatar_url} key={member.avatar_url} />
    })
    return (
      <div className="text-center fw-bold text-primary">
        {profiles}
        {remaining > 0 && `+${data.members.length - 5}`}
      </div>
    )
  }
  return (
    <div className="bg-white mt-3">
      <div className="container">
        <img src="/assets/discordLogo.svg" />
        <div className="text-primary mt-n3 text-center">
          <span className="fw-bold">{data.error || data.members.length}</span>
          {data.error
            ? ''
            : data.members.length === 1
              ? ' member online'
              : ' members online'}
        </div>
      </div>
      <div className="p-2">
        <div className={`${styles['discord-list']}`}>{reshuffle(data)}</div>
      </div>
      <a
        className={`${styles['discord_link']} fw-bold text-white`}
        href="https://discord.gg/c0d3"
        target="_blank"
        rel="noreferrer"
      >
        <div
          className={`border border-primary text-center bg-white text-primary`}
        >
          JOIN
        </div>
      </a>
    </div>
  )
}

export default DiscordBar
