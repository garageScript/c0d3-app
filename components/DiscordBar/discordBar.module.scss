.discord-list {
  max-height: 50px;
  min-height: 50px;
}
.discord_link {
  & > div {
    font-size: 1.3rem;
    border-bottom-left-radius: 5px;
    border-bottom-right-radius: 5px;
    letter-spacing: 0.1rem;
  }
  &:hover {
    text-decoration-color: rgb(84, 64, 216);
  }
}

.member-avatar {
  max-height: 50px;
  width: auto;
  border-radius: 50%;
  opacity: 80%;
}
