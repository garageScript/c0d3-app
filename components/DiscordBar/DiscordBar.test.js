import React from 'react'
import DiscordBar from './DiscordBar'
import { render, screen, waitFor } from '@testing-library/react'
describe('DiscordBar component', () => {
  const Bobby = {
    avatar_url:
      'https://cdn.discordapp.com/widget-avatars/Niju_ahWJDG-ehJlK3DLrBV2kceVq5gNEEb3pnDqO7Y/dpYnY4uiLVWlkMqjQCsG8vRqu9DUjZu_Burf1MHBQ2DxS0GUfACw7Or1oCLR061S9v906uMu3Wfl_e3HHizB1NfjbZfxiqHmAj2Ob7hM_xXARZvl2mgC7xq1y3jYKlgQQBGMVeQ_KutHDw',
    name: 'Bobby'
  }
  const Sly = {
    avatar_url:
      'https://cdn.discordapp.com/widget-avatars/JbWqF_-A-fVbRlrdu1vtB6JU1cQp8bNJIeoDfKt_5NM/suFvloPCkCf8iut738QMC2F6Pqvm02gd0m_fq7hHXtw15bmIDqmYQ-3xRhmScXEcjFGoJVB2Q-FZf37ClwvMP3t3JPfxELAdCpzSjq9GHM8X7bzaDuamjMZiOsHE6OuD7wlZf-i4FlrbBg',
    name: 'Sly'
  }
  const window = global.window
  const createMatcher = (number, string) => {
    return (_, node) => {
      return Array.from(node.children).every(
        child => child.textContent === number || child.textContent === string
      )
    }
  }
  beforeEach(() => {
    global.window = window
    global.Math.random = () => 0.6
  })
  test('Should render no data', async () => {
    global.window.fetch = jest.fn().mockResolvedValue({
      json: async () => ({})
    })

    render(<DiscordBar />)
    const matcher = createMatcher('0', 'members online')
    await waitFor(() => screen.getAllByText(matcher))
  })
  test('Should render one user', async () => {
    global.window.fetch = jest.fn().mockResolvedValue({
      json: async () => ({ presence_count: 0, members: [Bobby] })
    })
    const { container } = render(<DiscordBar />)
    const matcher = createMatcher('1', 'member online')
    await waitFor(() => screen.getAllByText(matcher))
    expect(container).toMatchSnapshot()
  })
  test('Should render two users', async () => {
    global.window.fetch = jest.fn().mockResolvedValue({
      json: async () => ({ presence_count: 2, members: [Bobby, Sly] })
    })
    const { container } = render(<DiscordBar />)
    const matcher = createMatcher('2', 'members online')
    await waitFor(() => screen.getAllByText(matcher))
    expect(container).toMatchSnapshot()
  })
  test('Should render twenty users', async () => {
    global.window.fetch = jest.fn().mockResolvedValue({
      json: async () => ({ presence_count: 20, members: Array(20).fill(Bobby) })
    })
    //disable warnings about 20 children with the same key, production keys are always unique
    jest.spyOn(console, 'error').mockImplementation()
    render(<DiscordBar />)
    const matcher = createMatcher('20', 'members online')
    await waitFor(() => screen.getAllByText(matcher))
  })
  test('Should restart fetching after error', () => {
    global.window.fetch = jest.fn().mockRejectedValue({
      json: () => {}
    })
    render(<DiscordBar />)
  })
})
