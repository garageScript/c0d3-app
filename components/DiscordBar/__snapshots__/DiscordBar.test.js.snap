// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiscordBar component Should render one user 1`] = `
<div>
  <div
    class="bg-white mt-3"
  >
    <div
      class="container"
    >
      <img
        src="/assets/discordLogo.svg"
      />
      <div
        class="text-primary mt-n3 text-center"
      >
        <span
          class="fw-bold"
        >
          1
        </span>
         member online
      </div>
    </div>
    <div
      class="p-2"
    >
      <div
        class="discord-list"
      >
        <div
          class="text-center fw-bold text-primary"
        >
          <div
            class="d-inline-block"
          >
            <img
              class="member-avatar mx-1"
              src="https://cdn.discordapp.com/widget-avatars/Niju_ahWJDG-ehJlK3DLrBV2kceVq5gNEEb3pnDqO7Y/dpYnY4uiLVWlkMqjQCsG8vRqu9DUjZu_Burf1MHBQ2DxS0GUfACw7Or1oCLR061S9v906uMu3Wfl_e3HHizB1NfjbZfxiqHmAj2Ob7hM_xXARZvl2mgC7xq1y3jYKlgQQBGMVeQ_KutHDw"
            />
          </div>
        </div>
      </div>
    </div>
    <a
      class="discord_link fw-bold text-white"
      href="https://discord.gg/c0d3"
      rel="noreferrer"
      target="_blank"
    >
      <div
        class="border border-primary text-center bg-white text-primary"
      >
        JOIN
      </div>
    </a>
  </div>
</div>
`;

exports[`DiscordBar component Should render two users 1`] = `
<div>
  <div
    class="bg-white mt-3"
  >
    <div
      class="container"
    >
      <img
        src="/assets/discordLogo.svg"
      />
      <div
        class="text-primary mt-n3 text-center"
      >
        <span
          class="fw-bold"
        >
          2
        </span>
         members online
      </div>
    </div>
    <div
      class="p-2"
    >
      <div
        class="discord-list"
      >
        <div
          class="text-center fw-bold text-primary"
        >
          <div
            class="d-inline-block"
          >
            <img
              class="member-avatar mx-1"
              src="https://cdn.discordapp.com/widget-avatars/JbWqF_-A-fVbRlrdu1vtB6JU1cQp8bNJIeoDfKt_5NM/suFvloPCkCf8iut738QMC2F6Pqvm02gd0m_fq7hHXtw15bmIDqmYQ-3xRhmScXEcjFGoJVB2Q-FZf37ClwvMP3t3JPfxELAdCpzSjq9GHM8X7bzaDuamjMZiOsHE6OuD7wlZf-i4FlrbBg"
            />
          </div>
          <div
            class="d-inline-block"
          >
            <img
              class="member-avatar mx-1"
              src="https://cdn.discordapp.com/widget-avatars/Niju_ahWJDG-ehJlK3DLrBV2kceVq5gNEEb3pnDqO7Y/dpYnY4uiLVWlkMqjQCsG8vRqu9DUjZu_Burf1MHBQ2DxS0GUfACw7Or1oCLR061S9v906uMu3Wfl_e3HHizB1NfjbZfxiqHmAj2Ob7hM_xXARZvl2mgC7xq1y3jYKlgQQBGMVeQ_KutHDw"
            />
          </div>
        </div>
      </div>
    </div>
    <a
      class="discord_link fw-bold text-white"
      href="https://discord.gg/c0d3"
      rel="noreferrer"
      target="_blank"
    >
      <div
        class="border border-primary text-center bg-white text-primary"
      >
        JOIN
      </div>
    </a>
  </div>
</div>
`;
